### DISCUSSION TITLE
any easy way to revert last `do`, or rather `undo` last action?

### DISCUSSION THREAD

**Discussion by vt128** [todotxt/todo.txt-cli#324] [Q&A] [answered]
It seems that `todo.txt.bak` can be helpful to revert last action on `todo.txt`, but `done.txt` doesn't have such thing.

so can you please let me know if there is such a plugin or verb in future?

**Comment by inkarkat** [comment, marked as answer]
`todo.txt` is modified by many actions, and in various ways (add tasks, delete tasks, modify parts of a task). Internally, these modifications are done via `sed`, and that creates the `todo.txt.bak` file.

For `done.txt`, it is only ever appended to, via the `archive` action (or auto-archiving if configured). So undoing here means removing task(s) from the end and moving them back to `todo.txt`. You don't need a backup file to do this correctly; a simple text editor (or `sed`) will do.

Unless you enable auto-archiving, the main undo procedure for accidentally done tasks is entirely in `todo.txt`; I usually would not use the `.txt.bak` file for that, but rather use the built-in `replace` action or a custom add-on like my [undo](https://github.com/inkarkat/todo.txt-cli-ex/blob/master/actions/undo). Archiving typically is only run occasionally to clean up; I do this once a week.

External tools, either from the operating system (time machine) or Dropbox can also help you recovering earlier versions of any todo.sh data file.

**Comment by vt128** [comment]
Thanks for your details!

Auto-archiving is enabled by default, I just turned it off via `export TODOTXT_AUTO_ARCHIVE=0`. Everything works just fine now ---
1) use `edit` plugin to edit last line with vim;
2) my own `undo` plugin based on `git reset --hard` --- however i need to commit more often;
